plat/common/include/arm/arm64/paging.h

```diff
@@ -0,0 +1,381 @@
+
+	switch (attr & (PAGE_ATTR_SHAREABLE_MASK <<
+		PAGE_ATTR_SHAREABLE_SHIFT)) {
+	case PAGE_ATTR_SHAREABLE_IS:
+		pte |= PTE_ATTR_SH(PTE_ATTR_SH_IS);
+		break;
+	case PAGE_ATTR_SHAREABLE_OS:
+		pte |= PTE_ATTR_SH(PTE_ATTR_SH_OS);
+		break;
+	case PAGE_ATTR_SHAREABLE_NS:
+		pte |= PTE_ATTR_SH(PTE_ATTR_SH_NS);
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid shareability type\n");
+	}
+
+	switch (attr & (PAGE_ATTR_TYPE_MASK << PAGE_ATTR_TYPE_SHIFT)) {
+	case PAGE_ATTR_TYPE_NORMAL_WB:
+		pte |= PTE_ATTR_IDX(NORMAL_WB);
+		break;
+	case PAGE_ATTR_TYPE_NORMAL_WT:
+		pte |= PTE_ATTR_IDX(NORMAL_WT);
+		break;
+	case PAGE_ATTR_TYPE_NORMAL_NC:
+		pte |= PTE_ATTR_IDX(NORMAL_NC);
+		break;
+	case PAGE_ATTR_TYPE_DEVICE_nGnRnE:
+		pte |= PTE_ATTR_IDX(DEVICE_nGnRnE);
+		break;
+	case PAGE_ATTR_TYPE_DEVICE_nGnRE:
+		pte |= PTE_ATTR_IDX(DEVICE_nGnRE);
+		break;
+	case PAGE_ATTR_TYPE_DEVICE_GRE:
+		pte |= PTE_ATTR_IDX(DEVICE_GRE);
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid memory type\n");
+	}
+
+	return pte;
+}
+
+static inline __pte_t
+pgarch_pte_change_attr(__pte_t pte, unsigned long new_attr,
+		       unsigned int level __unused)
+{
+	(void)pte;
+	(void)new_attr;
+
+	UK_CRASH("%s: Not implemented", __func__);
+
+	return 0;
+}
+
+static inline unsigned long
+pgarch_attr_from_pte(__pte_t pte, unsigned int level __unused)
+{
+	unsigned long attr = PAGE_ATTR_PROT_READ;
+
+	if (!(pte & PTE_ATTR_AP(PTE_ATTR_AP_RO)))
+		attr |= PAGE_ATTR_PROT_WRITE;
+
+	if (!(pte & PTE_ATTR_PXN))
+		attr |= PAGE_ATTR_PROT_EXEC;
+
+	switch (pte & PTE_ATTR_SH_MASK) {
+	case (PTE_ATTR_SH(PTE_ATTR_SH_NS)):
+		attr |= PAGE_ATTR_SHAREABLE_NS;
+		break;
+	case (PTE_ATTR_SH(PTE_ATTR_SH_IS)):
+		attr |= PAGE_ATTR_SHAREABLE_IS;
+		break;
+	case (PTE_ATTR_SH(PTE_ATTR_SH_OS)):
+		attr |= PAGE_ATTR_SHAREABLE_OS;
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid shareability type\n");
+	};
+
+	switch (pte & PTE_ATTR_IDX_MASK) {
+	case PTE_ATTR_IDX(NORMAL_WB):
+		attr |= PAGE_ATTR_TYPE_NORMAL_WB;
+		break;
+	case PTE_ATTR_IDX(NORMAL_WT):
+		attr |= PAGE_ATTR_TYPE_NORMAL_WT;
+		break;
+	case PTE_ATTR_IDX(NORMAL_NC):
+		attr |= PAGE_ATTR_TYPE_NORMAL_NC;
+		break;
+	case PTE_ATTR_IDX(DEVICE_nGnRnE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_nGnRnE;
+		break;
+	case PTE_ATTR_IDX(DEVICE_nGnRE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_nGnRE;
+		break;
+	case PTE_ATTR_IDX(DEVICE_GRE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_GRE;
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid memory type\n");
+	}
+
+	return attr;
+}
+
+/*
+ * Page tables
+ */
+static inline __vaddr_t
+pgarch_pt_pte_to_vaddr(struct uk_pagetable *pt __unused, __pte_t pte,
+		       unsigned int level __maybe_unused)
+{
+	(void)pt;
+
+	return arm64_directmap_paddr_to_vaddr(PT_Lx_PTE_PADDR(pte, level));
+}
+
+/* Create a table descriptor */
+static inline __pte_t
+pgarch_pt_pte_create(struct uk_pagetable *pt __unused, __paddr_t pt_paddr,
+		     unsigned int level __unused, __pte_t template,
+		     unsigned int template_level __unused)
+{
+	(void)template;
+
```

Comment:
Correct. That is because I haven't implemented yet support for next-level attributes of table descriptors. For now just replacing `(void)templat` with `__unused`.

Reply:
Maybe add a `TODO` comment?